### ISSUE TITLE
[FEAT]: Repository Rulesets

### ISSUE THREAD

**Issue by rmvangun** [integrations/terraform-provider-github#1777]
### Describe the need

Now that the Repository Ruleset feature is available to Enterprise customers, we would like to be able to codify these rules via Terraform. Here the blog post on repository rules: https://github.blog/changelog/2023-04-17-introducing-repository-rules-public-beta/

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kfcampbell** [comment]
I don't believe the API for this feature has been released; can you find documentation on it?

**Comment by YElyousfi** [comment]
> I don't believe the API for this feature has been released; can you find documentation on it?

Both organization and repository rulesets have been added in this release: https://github.com/google/go-github/releases/tag/v53.2.0

**Comment by kfcampbell** [comment]
Awesome news, thank you @YElyousfi!

**Comment by felixlut** [comment]
I can take a stab at this if no one else has started yet

**Comment by goffinf** [comment]
@felixlut  thank you for your swift response. I may have conflated the version of the Go library release tag with this work, but I really appreciate your effort.

I am heading up a project that is rolling out EMU, GitHub Actions and GitHub Advanced Security which is where my interest comes from. I can reach out to our CSM about potential changes to the GA release of Rulesets if it would help but I note you have already done so to @kfcampbell.

On the topic of automated config, right now we have to use the GitHub API (via Octokit) to create Org level rulesets, but, whilst I don’t mind the programmatic approach, I would prefer using Terraform for all of the obvious reasons. I have heard others say that Terraform can be problematic (rate limit issues and some state problems, for example).

@kfcampbell do you have a view on whether Octokit is the recommended approach vs. Terraform ?

**Comment by YElyousfi** [comment]
Are **_org_** rulesets being covered by this issue as well?

**Comment by felixlut** [comment]
I've created a Draft PR for this issues now! I'd like some feedback on the structure of the Terraform schema, see #1807. 

This feature is also blocked until [`go-github`](https://github.com/google/go-github) creates a new release. The Ruleset API seems to have made some minor changes since their last one (it's a beta feature after all 😓). @goffinf if you could reach out to you CSM that would be great! If you could, I'd like some clarifications on the comment [I've made here](https://github.com/integrations/terraform-provider-github/pull/1807#discussion_r1271416422), because the API seems a bit inconsistent there.

@YElyousfi I've only implemented Repository level ones so far, but I don't thing the Org ones should required that much work tbh. Repository and Org level Rules are very similar. so once my questionmarks in #1807 has been resolved, I could look at the Org level ones as well. I'm more interested in those anyways, it just make more sense to implement the Repository level one first 😄

**Comment by felixlut** [comment]
> @felixlut thank you for your swift response. I may have conflated the version of the Go library release tag with this work, but I really appreciate your effort.
> 
> I am involved in a project that is rolling out EMU, GitHub Actions and GitHub Advanced Security which is where my interest comes from. I can reach out to our CSM about potential changes to the GA release of Rulesets if it would help but I note you have already done so to @kfcampbell.
> 
> On the topic of automated config, right now we have to use the GitHub API (via Octokit) to create Org level rulesets, but, whilst I don’t mind the programmatic approach, I would prefer using Terraform for all of the obvious reasons. I have heard others say that Terraform can be problematic (rate limit issues and some state problems, for example).
> 
> @kfcampbell do you have a view on whether Octokit is the recommended approach vs. Terraform ?

I'm not affiliated in any way with GitHub, so take this for what it is; the opinion of someone who have been using Terraform to mange GitHub ([SonarQube](https://registry.terraform.io/providers/jdamata/sonarqube/latest/docs), [AzureAD](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs), [Azure Cloud](https://registry.terraform.io/providers/hashicorp/azurerm/latest)) extensively for the last half year or so as a part of the `DevOps` team. 

I'm a big fan of the [`GitOps`](https://opengitops.dev/) principles, and declarative over imperative solutions (the configuration in your repos `main` branch is then also applied to the real world). I would take a Terraform solution over a scripted one any day of the week, I think that is the more elegant and maintainable approach. I've ran into cases where I run `terraform apply` followed by a few direct calls to the API to fix stuff that isn't yet implemented in the provider. That is never a good feeling. Solving everything the same way, within the same "box" is very useful. 

I see huge value in having 1 common interface (Terraform in this case) to solve problems like this. If we are multiple developers in my team, and we all have Terraform as a common language that we all know how to interact with, that is really useful. Then it doesn't matter what platform we control via Terraform, all members have a "foot in the door" so to speak, because they can read the Terraform, and then map it to the platform in question with little to no previous knowledge of that platform. Integrating the different platforms together is also easier if you have a common solution like Terraform. Eg., maybe you want each [`github_team`](https://registry.terraform.io/providers/integrations/github/latest/docs/data-sources/team) to also have an [`azurerm_resource_group`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_group.html) and [`key_vault`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault) in Azure. And permissions for those? How about each `github_team` is synched to an [azuread_group](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/resources/group) with [`github_team_sync_group_mapping`](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/team_sync_group_mapping), and the permissions to the `resource_group` and `key_vault` in Azure is automatically bootstrapped at the same time. Want to add a new platform, lets say Launch Darkly, to the mix? Just map each `github_team` to a [`launchdarkly_team`](https://registry.terraform.io/providers/launchdarkly/launchdarkly/latest/docs/resources/team), and give them the proper permissions with [`team_role_mapping`](https://registry.terraform.io/providers/launchdarkly/launchdarkly/latest/docs/resources/team_role_mapping) over their own [`launchdarkly_project`](https://registry.terraform.io/providers/launchdarkly/launchdarkly/latest/docs/resources/project). All of this wouldn't be more than a couple of hundred lines of Terraform. Of course it won't be that easy, and edgecases shows up, but this idea is pretty powerful!

Terraform is far from perfect though, and we've ran into stuff like rate limiting as you mention. A big reason for this is that, at least from what I've seen, it pretty common for an entire organization have a huge mono-terraform-repo which all developers make changes to. While this works, it's not really ideal to have a single repo that hundred, or even thousands of people contribute to (pull request sprawling, who "owns" what, who should approve which PR, API rate limiting, among others). Splitting up Terraform into smaller repos, where each team owns their stuff, with a centralized repo containing Terraform modules and shared code, is not trivial either. Each repo needs to be applied, have an s3 bucket for the statefile, permissions in whatever platform that bucket is located in. It's not trivial. 

My conclusion has been that there is no issue to run Terraform for my own stuff, but the second another team is involved and needs to create resources, Terraform doesn't scale that well. It's very good at what it does, but it's very rigid. My team have started looking into [Crossplane](https://www.crossplane.io/) (in a nutshell its like terraform, but you have kubernetes yamls instead of terraform code) as a solution for allowing our developers to create their own resources. It looks pretty promising, but we haven't gotten much farther than some reading/videos. A [Crossplane GitHub provider](https://github.com/crossplane-contrib/provider-github) does exist, as well as [Terraform one](https://github.com/upbound/provider-terraform#official-terraform-provider) (which essentially allow you to deploy Terraform as a kubernetes yaml, and the cluster reconciles it). 

That became way longer than I thought it would be 😓. Been meaning to write this down as findings internally in my teams for a while anyways, so thanks for giving me a reason to clear my thoughts a bit!

**Comment by goffinf** [comment]
@felixlut I have raised this with my CSM at GitHub but haven’t heard anything yet (I will chase up but I know a lot of them are out on a conference so I don’t expect to get a fast response unless its an urgent issue).

On the topic of automation options for GitHub configuration I have looked at 6 in as unbiased a way as I can over the past few weeks. I can say openly I am a big fan of Terraform for the declarative desired state convergence model as well as how the abstractions deal with the lifecycle of the referenced resources. Since the GitHub provider doesn’t cover Org and Repo level rulesets and I want them to be managed just like any other GH resource as part of on-boarding I have had to look elsewhere. I do have a working implementation using JavaScript using the Octokit REST client that can be executed within a GitHub Actions workflow using the github-script action. That’s all well and good, but, … the amount of coding effort required to provide the same level of idempotence guarantees is night-and-day different from the declarative approach. Not only that, you find yourself inventing a custom DSL so that you implement a configuration driven approach. It’s not unsustainable, especially if you have a dedicated DevSecOps or SCM platform team to look after the code base and maintain it as the GH API evolves, but TBH its a LOT more complexity than I am comfortable with. That said, for a particular set of use cases I am aware of some issues around rate limiting (apparently the Terraform provider can be quite ‘chatty’ for some resource types), state management and access controls.

It is probably inevitable that there will be a need for a hybrid approach unless you go all-in on lower level scripting.

However, since my team isn’t responsible for the whole of the GH platform I would be much happier using Terraform at least from a maintenance perspective and to leverage a common skill set.

So, I am very supportive of the work you are doing on the provider. Like you I need Org level rulesets rather than Repo level.

If there is anything I can help with (testing for example) I would be very happy to, I have a bit of free time this coming week ?

Kind Regards

Fraser.

**Comment by felixlut** [comment]
@goffinf Thank you for that! 

I don't think there is much to do until google/go-github gets a bump, except looking at the [Terraform Schema](https://github.com/integrations/terraform-provider-github/pull/1807#issue-1817054708) to see that it fits your use-case. I know you said you're mostly here for the Org level ones, but they'd look very similar 😄

**Comment by kfcampbell** [comment]
The discussion here has been fascinating, thank you @felixlut and @goffinf; it's great to hear this kind of experience report. 

I'd like to respond to 

> @kfcampbell do you have a view on whether Octokit is the recommended approach vs. Terraform ?

I don't recommend one over the other. They both have strengths and weaknesses as you described above, and @nickfloyd and I do our best to support both communities.

**Comment by goffinf** [comment]
@kfcampbell / @nickfloyd wrt: automation options for GitHub resources … understood, good to hear both are proactively supported.

I confess that I am of a similar mind although I would prefer to coalesce around a single approach if I can.

For my own specific use case, which is related to a specific subset of GH resources (those concerned with configuring GitHub Advanced Security (GHAS) at the Org and repo levels - so things such as Org rulesets as are being discussed here but also webhooks and GHA workflow config and a few other things), Terraform is a good fit. I had intended to use Octokit as a tactical stop-gap until the provider supports Org rulesets, but actually, since I am expecting the shelf-life for this to be short (hoping @felixlut can get this change over the line in the coming weeks) I might go to the GitHub CLI and just feed it the request JSON from the HTTP API and use that as a ‘poor mans’ declaration of desired state for now so the code is at least somewhat flexible. The implementation seems simpler (I’m going to test consuming JSON (maybe YAML) over the next couple of days to drive a create or update of an Org ruleset based on whether it exists or not - which I can find out using the ruleset name and a simple gh api get).

That leads me onto the ruleset schema which @felixlut has been asking about above. Are either of you able to comment on the questions which Felix has raised around some changes to the API and whether any others are likely in the very near term ?

… and also when the new google/go-github version will be released ?

@felixlut … when I was playing around with Octokit and Org rulesets I did notice a couple of things that were slightly at odds with the docs. For example, the repository_name and the ref_name structure MUST include BOTH the include AND exclude arrays even if they are empty otherwise a 422 status code is returned (repository name can also include a protected property but that one is optional). I’m sure you are already aware but if not, it might have an impact on what you are doing and the TF schema ?

**Comment by felixlut** [comment]
> @kfcampbell / @nickfloyd wrt: automation options for GitHub resources … understood, good to hear both are proactively supported.
> 
> I confess that I am of a similar mind although I would prefer to coalesce around a single approach if I can.
> 
> For my own specific use case, which is related to a specific subset of GH resources (those concerned with configuring GitHub Advanced Security (GHAS) at the Org and repo levels - so things such as Org rulesets as are being discussed here but also webhooks and GHA workflow config and a few other things), Terraform is a good fit. I had intended to use Octokit as a tactical stop-gap until the provider supports Org rulesets, but actually, since I am expecting the shelf-life for this to be short (hoping @felixlut can get this change over the line in the coming weeks) I might go to the GitHub CLI and just feed it the request JSON from the HTTP API and use that as a ‘poor mans’ declaration of desired state for now so the code is at least somewhat flexible. The implementation seems simpler (I’m going to test consuming JSON (maybe YAML) over the next couple of days to drive a create or update of an Org ruleset based on whether it exists or not - which I can find out using the ruleset name and a simple gh api get).
> 
> That leads me onto the ruleset schema which @felixlut has been asking about above. Are either of you able to comment on the questions which Felix has raised around some changes to the API and whether any others are likely in the very near term ?
> 
> … and also when the new google/go-github version will be released ?
> 
> @felixlut … when I was playing around with Octokit and Org rulesets I did notice a couple of things that were slightly at odds with the docs. For example, the repository_name and the ref_name structure MUST include BOTH the include AND exclude arrays even if they are empty otherwise a 422 status code is returned (repository name can also include a protected property but that one is optional). I’m sure you are already aware but if not, it might have an impact on what you are doing and the TF schema ?

@goffinf I'm aware of that specific gotcha, but I'm sure there are more that I've missed. That's what tests are for 😅 

So latest update on this feature is that me and @o-sama have started collaborating on this after noticing that we we're doing the same work in parallel. We've consolidated the "best" parts of each implementation in #1808, and gotten org rules up and running as well. Still some ironing out to do (tests being the big one), and still waiting on a bump for `go-github` (along with a resolution to the bug in https://github.com/google/go-github/pull/2836), but we are progressing 😄

**Comment by stevehipwell** [comment]
FYI repository rules are now [GA](https://github.blog/2023-07-24-github-repository-rules-are-now-generally-available/).

**Comment by goffinf** [comment]
Woohoo, great work. Dare I ask about Org level rulesets ?

**Comment by o-sama** [comment]
@felixlut and I already got org-level rule sets mostly working in #1808 and my next commit should get them fully functional. Of course we still have to write tests and all but I have tested manually and it's looking good!

**Comment by kfcampbell** [comment]
> Are either of you able to comment on the questions which Felix has raised around some changes to the API and whether any others are likely in the very near term ?

I don't have any special internal exposure to the API team's backlog unfortunately; I wish I could offer anything more. 

> … and also when the new google/go-github version will be released ?

It's always been weird to me that some GitHub projects like this depend heavily on the google/go-github library, which we don't own or have any influence over. I'm hopeful that in the future we can provide our own SDKs generated from GitHub's OpenAPI schema so we'll be able to own the entire library stack that comprises the provider. In the meantime, `@gmlewis` has been responsive on that repo and may be able to help in an issue or discussion over there. 

As a side note: I love to see the collaboration and activity on this repo! It makes me happy that people are finding it worthwhile to contribute and to see the community so active.

**Comment by o-sama** [comment]
Update: @felixlut and I have gotten the feature to a good state at the moment, we’re still testing it out and trying to iron out anything we find, but the main blocker is the next [go-github](https://github.com/google/go-github) release, which will include a couple of fixes that are required to use rulesets.

**Comment by goffinf** [comment]
That’s great to hear, have you also brought the Org level rulesets into play ?

I have coded ruleset creation using the gh api … but I would obviously prefer to bring that into TF also.

Any idea when the go-github release will land ?

**Comment by o-sama** [comment]
Yes, we have! I’ve successfully tested all CRUD operations on organization rulesets manually via provider dev overrides, but automated testing will have to be done using a test enterprise which I don’t have access to.

My understanding is that the go-github release will be soon since I saw a mention of merging one more PR before cutting it, but I don’t have a timeline, my best estimate is within a week but I could very well be wrong 😅

**Comment by goffinf** [comment]
I can help with testing within an Enterprise account. I am an Org owner of a sandbox org which we use specifically for testing GitHub automation.

How can I help ?

**Comment by pcassaretto-cmc** [comment]
https://github.blog/changelog/2023-08-02-github-actions-required-workflows-will-move-to-repository-rules/

Unfortunately it seems there's a deadline for this functionality to allow required workflows to be deployed in rulesets via tf, and required checks in branch protection rules may no longer work from Oct 18th.  Fully appreciate the efforts of adding the necessary resources to the provider.

**Comment by o-sama** [comment]
@pcassaretto-cmc yeah it looks like we'll have to revisit rulesets and add required workflows at some point between its release and the deadline, but should be good for now since we're about 50 days away from release. 

I didn't see anything about required status checks leaving rulesets though in the article, was that mentioned elsewhere?

**Comment by pcassaretto-cmc** [comment]
> I didn't see anything about required status checks leaving rulesets though in the article, was that mentioned elsewhere?

@o-sama this is mentioned in the post:

> users will no longer be able to access Actions Required Workflows and must use rulesets in its place.

I assumed this meant the required checks in branch protection rules are being removed - I think that's the only place required checks exist at the moment?

**Comment by o-sama** [comment]
AFAIK required checks != required workflows. Checks would be a pipeline run passing or a TF Cloud speculative plan being successful for example, workflows on the other hand would be GitHub Actions, for example running a linter or unit tests.

**Comment by goffinfnbs** [comment]
I am reaching out to my GH CSM to ask about whether there is any change to required status checks within rulesets and how the required workflow feature will be represented. I'll post back when I have a response.

**Comment by goffinf** [comment]
I ask my CSM at GitHub at every bi-weekly meeting but as yet they haven’t provided anything definitive other than to say that required workflows will be absorbed into the rulesets structure but that they dont have full details yet (surprising since the deprecation of required workflows is close and some orgs may have a bunch of existing definitions to move).

That aside, has there been any further progress possible here with the the go GitHub release or anything else ?

**Comment by o-sama** [comment]
`go-github` has been upgraded here, the PR is just awaiting a review, merge, and then a release.
